// @renproject/chains/build/module/bch.js
import { isMainnetAddress, isTestnetAddress, toCashAddress } from "bchaddrjs";
import { Networks, Opcode, Script } from "bitcore-lib-cash";
import { getConfirmations, getUTXO, getUTXOs, } from "send-crypto/build/main/handlers/BCH/BCHHandler";
import { anyAddressFrom } from "./btc";
import { createAddress } from "./common";
import { Ox, strip0x } from "./hex";
export const createBCHAddress = createAddress(Networks, Opcode, Script);
export const getBitcoinCashUTXOs = ({ isTestnet }) => {
    return async (address, confirmations) => {
        return getUTXOs(isTestnet, { address, confirmations });
    };
};
export const getBitcoinCashConfirmations = ({ isTestnet, }) => {
    return async (txHash) => {
        return getConfirmations(isTestnet, txHash);
    };
};
export const getBitcoinCashUTXO = ({ isTestnet }) => {
    return async (txHash, vOut) => {
        return getUTXO(isTestnet, txHash, vOut);
    };
};
export const bchAddressToHex = (address) => Ox(Buffer.from(address));
const isBCHAddress = (address, options) => {
    try {
        return options
            ? options.isTestnet
                ? isTestnetAddress(address)
                : isMainnetAddress(address)
            : isTestnetAddress(address) || isMainnetAddress(address);
    }
    catch (error) {
        return false;
    }
};
const bchTactics = {
    decoders: [
        (address) => Buffer.from(address),
        (address) => Buffer.from(address, "base64"),
        (address) => Buffer.from(strip0x(address), "hex"),
    ],
    encoders: [(buffer) => toCashAddress(buffer.toString())],
};
export const bchAddressFrom = anyAddressFrom(isBCHAddress, bchTactics);
