// @renproject/chains/build/module/btc.js
import { Networks, Opcode, Script } from "bitcore-lib";
import { encode } from "bs58";
import { getConfirmations, getUTXO, getUTXOs, } from "send-crypto/build/main/handlers/BTC/BTCHandler";
import { validate } from "wallet-address-validator";
import { createAddress } from "./common";
import { Ox, strip0x } from "./hex";
export const createBTCAddress = createAddress(Networks, Opcode, Script);
export const getBitcoinUTXOs = ({ isTestnet }) => {
    return async (address, confirmations) => {
        return getUTXOs(isTestnet, { address, confirmations });
    };
};
export const getBitcoinConfirmations = ({ isTestnet, }) => {
    return async (txHash) => {
        return getConfirmations(isTestnet, txHash);
    };
};
export const getBitcoinUTXO = ({ isTestnet }) => {
    return async (txHash, vOut) => {
        return getUTXO(isTestnet, txHash, vOut);
    };
};
// export const btcAddressToHex = (address: string) => {
//     const addressBuffer = new Address(address).toBuffer();
//     // Concatenate checksum
//     return Ox(Buffer.concat([addressBuffer, Base58Check.checksum(addressBuffer)]));
// };
// export const btcAddressFrom = (address: string, encoding: "hex" | "base64") => {
//     // tslint:disable-next-line: no-any
//     return (Address as any)
//         .fromBuffer(Buffer.from(encoding === "hex" ? strip0x(address) : address, encoding).slice(0, -4))
//         .toString();
// };
export const btcAddressToHex = (address) => Ox(Buffer.from(address));
const isBTCAddress = (address) => validate(address, "btc", "testnet") || validate(address, "btc", "prod");
const btcTactics = {
    decoders: [
        (address) => Buffer.from(address),
        (address) => Buffer.from(address, "base64"),
        (address) => Buffer.from(strip0x(address), "hex"),
    ],
    encoders: [
        (buffer) => encode(buffer),
        (buffer) => buffer.toString(),
    ],
};
export const anyAddressFrom = (isAnyAddress, { encoders, decoders }) => (address) => {
    for (const encoder of encoders) {
        for (const decoder of decoders) {
            try {
                const encoded = encoder(decoder(address));
                if (isAnyAddress(encoded)) {
                    return encoded;
                }
            }
            catch (error) {
                // Ignore errors
            }
        }
    }
    return address;
};
export const btcAddressFrom = anyAddressFrom(isBTCAddress, btcTactics);
