// @renproject/chains/build/module/zec.js
import { Address, Networks, Opcode, Script } from "bitcore-lib-zcash";
import Base58Check from "bitcore-lib-zcash/lib/encoding/base58check";
import { encode } from "bs58";
import { getConfirmations, getUTXO, getUTXOs, } from "send-crypto/build/main/handlers/ZEC/ZECHandler";
import { validate } from "wallet-address-validator";
import { anyAddressFrom } from "./btc";
import { createAddress } from "./common";
import { Ox, strip0x } from "./hex";
export const createZECAddress = createAddress(Networks, Opcode, Script);
export const getZcashUTXOs = ({ isTestnet }) => {
    return async (address, confirmations) => {
        return getUTXOs(isTestnet, { address, confirmations });
    };
};
export const getZcashConfirmations = ({ isTestnet, }) => {
    return async (txHash) => {
        return getConfirmations(isTestnet, txHash);
    };
};
export const getZcashUTXO = ({ isTestnet }) => {
    return async (txHash, vOut) => {
        return getUTXO(isTestnet, txHash, vOut);
    };
};
// ZCash shielded addresses (starting with 'z') aren't supported yet.
// export const zecAddressToHex = (address: string) => Ox(Buffer.from(address));
export const zecAddressToHex = (address) => {
    const addressBuffer = new Address(address).toBuffer();
    // Concatenate checksum
    return Ox(Buffer.concat([addressBuffer, Base58Check.checksum(addressBuffer)]));
};
const isZECAddress = (address) => validate(address, "zec", "testnet") || validate(address, "zec", "prod");
const zecTactics = {
    decoders: [
        (address) => Buffer.from(address),
        (address) => Buffer.from(address, "base64"),
        (address) => Buffer.from(strip0x(address), "hex"),
    ],
    encoders: [
        (buffer) => buffer.toString(),
        (buffer) => encode(buffer),
    ],
};
export const zecAddressFrom = anyAddressFrom(isZECAddress, zecTactics);
