// @renproject/gateway/build/main/utils.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@renproject/utils");
// For now, the endpoints are network specific.
exports.GATEWAY_ENDPOINT_STAGING = "https://gateway-staging.renproject.io/";
exports.GATEWAY_ENDPOINT_PRODUCTION = "https://renproject.github.io/gateway-staging.renproject.io/";
exports.getElement = (id) => {
    const element = document.getElementById(id);
    if (!element) {
        throw new Error(`Unable to find element ${id}`);
    }
    return element;
};
exports.createElementFromHTML = (htmlString) => {
    const div = document.createElement("div");
    // tslint:disable-next-line: no-object-mutation no-inner-html
    div.innerHTML = htmlString.trim();
    return div.firstChild;
};
exports.resolveEndpoint = (endpointIn, network, pathIn, transferID) => {
    // Remove ending '/' from endpoint
    const endpoint = endpointIn.slice(endpointIn.length - 1) === "/"
        ? endpointIn.slice(0, endpointIn.length - 1)
        : endpointIn;
    // Remove starting '/' from path
    const path = pathIn.slice(0, 1) === "/" ? pathIn.slice(1, pathIn.length) : pathIn;
    return `${endpoint}/#/${path}?network=${network.name}&${transferID ? `id=${transferID}` : ""}`;
};
// tslint:disable-next-line: readonly-keyword no-any
const fixBigNumber = (value, key) => {
    try {
        // tslint:disable-next-line: strict-type-predicates
        if (value[key] && typeof value[key] === "object") {
            // tslint:disable-next-line: no-object-mutation no-any
            value[key] = utils_1.toFixed(value[key]);
        }
    }
    catch (error) {
        // Ignore error - may be readonly value
    }
};
/**
 * prepareParamsForSendMessage turns possible BigNumber values into strings
 * before passing the params to sendMessage.
 * The error message `can't clone ...` is thrown if this step is skipped.
 * @param transferParamsIn The parameters being fixed.
 */
exports.prepareParamsForSendMessage = (transferParamsIn) => {
    // Certain types can't be sent via sendMessage - e.g. BigNumbers.
    const _a = transferParamsIn, { web3Provider } = _a, transferParamsFiltered = __rest(_a, ["web3Provider"]);
    const transferParams = transferParamsFiltered;
    // tslint:disable-next-line: no-unnecessary-type-assertion
    fixBigNumber(transferParams, "suggestedAmount");
    // Contract call values
    try {
        // tslint:disable-next-line: no-unnecessary-type-assertion
        const contractCalls = transferParams
            .contractCalls;
        if (contractCalls) {
            for (const contractCall of contractCalls) {
                const contractParamsInner = contractCall.contractParams;
                if (contractParamsInner) {
                    for (const contractParam of contractParamsInner) {
                        fixBigNumber(contractParam, "value");
                    }
                }
            }
        }
        // tslint:disable-next-line: no-console
    }
    catch (error) {
        console.error(error);
    }
    return transferParams;
};
