// @renproject/interfaces/build/module/renVM.js
export var RenNetwork;
(function (RenNetwork) {
    RenNetwork["Mainnet"] = "mainnet";
    RenNetwork["Chaosnet"] = "chaosnet";
    RenNetwork["Testnet"] = "testnet";
    RenNetwork["Devnet"] = "devnet";
    RenNetwork["Localnet"] = "localnet";
})(RenNetwork || (RenNetwork = {}));
export const RenNetworks = [RenNetwork.Mainnet, RenNetwork.Chaosnet, RenNetwork.Testnet, RenNetwork.Devnet, RenNetwork.Localnet];
export const isRenNetwork = (maybeRenNetwork) => // tslint:disable-line: no-any
 RenNetworks.indexOf(maybeRenNetwork) !== -1;
export var Chain;
(function (Chain) {
    Chain["Bitcoin"] = "Btc";
    Chain["Ethereum"] = "Eth";
    Chain["Zcash"] = "Zec";
    Chain["BitcoinCash"] = "Bch";
})(Chain || (Chain = {}));
export const Chains = [Chain.Bitcoin, Chain.Ethereum, Chain.Zcash, Chain.BitcoinCash];
export const isChain = (maybeChain) => // tslint:disable-line: no-any
 Chains.indexOf(maybeChain) !== -1;
export var Asset;
(function (Asset) {
    Asset["BTC"] = "BTC";
    Asset["ZEC"] = "ZEC";
    Asset["ETH"] = "ETH";
    Asset["BCH"] = "BCH";
})(Asset || (Asset = {}));
export const Assets = [Asset.BTC, Asset.ZEC, Asset.ETH, Asset.BCH];
export const isAsset = (maybeAsset) => // tslint:disable-line: no-any
 Assets.indexOf(maybeAsset) !== -1;
export var RenContract;
(function (RenContract) {
    RenContract["Btc2Eth"] = "BTC0Btc2Eth";
    RenContract["Eth2Btc"] = "BTC0Eth2Btc";
    RenContract["Zec2Eth"] = "ZEC0Zec2Eth";
    RenContract["Eth2Zec"] = "ZEC0Eth2Zec";
    RenContract["Bch2Eth"] = "BCH0Bch2Eth";
    RenContract["Eth2Bch"] = "BCH0Eth2Bch";
})(RenContract || (RenContract = {}));
export const RenContracts = [RenContract.Btc2Eth, RenContract.Eth2Btc, RenContract.Zec2Eth, RenContract.Eth2Zec, RenContract.Bch2Eth, RenContract.Eth2Bch];
export const isRenContract = (maybeRenContract) => // tslint:disable-line: no-any
 RenContracts.indexOf(maybeRenContract) !== -1;
export const Tokens = {
    // Bitcoin
    BTC: {
        Mint: RenContract.Btc2Eth,
        Btc2Eth: RenContract.Btc2Eth,
        Burn: RenContract.Eth2Btc,
        Eth2Btc: RenContract.Eth2Btc,
    },
    // Zcash
    ZEC: {
        Mint: RenContract.Zec2Eth,
        Zec2Eth: RenContract.Zec2Eth,
        Burn: RenContract.Eth2Zec,
        Eth2Zec: RenContract.Eth2Zec,
    },
    // Bitcoin Cash
    BCH: {
        Mint: RenContract.Bch2Eth,
        Bch2Eth: RenContract.Bch2Eth,
        Burn: RenContract.Eth2Bch,
        Eth2Bch: RenContract.Eth2Bch,
    },
};
export var RenVMType;
(function (RenVMType) {
    RenVMType["TypeAddress"] = "address";
    RenVMType["TypeStr"] = "str";
    RenVMType["TypeB32"] = "b32";
    RenVMType["TypeB"] = "b";
    RenVMType["TypeI8"] = "i8";
    RenVMType["TypeI16"] = "i16";
    RenVMType["TypeI32"] = "i32";
    RenVMType["TypeI64"] = "i64";
    RenVMType["TypeI128"] = "i128";
    RenVMType["TypeI256"] = "i256";
    RenVMType["TypeU8"] = "u8";
    RenVMType["TypeU16"] = "u16";
    RenVMType["TypeU32"] = "u32";
    RenVMType["TypeU64"] = "u64";
    RenVMType["TypeU128"] = "u128";
    RenVMType["TypeU256"] = "u256";
    RenVMType["TypeRecord"] = "record";
    RenVMType["TypeList"] = "list";
    // Ext
    RenVMType["ExtTypeEthCompatAddress"] = "ext_ethCompatAddress";
    RenVMType["ExtTypeBtcCompatUTXO"] = "ext_btcCompatUTXO";
    RenVMType["ExtTypeBtcCompatUTXOs"] = "ext_btcCompatUTXOs";
    RenVMType["ExtTypeEthCompatTx"] = "ext_ethCompatTx";
    RenVMType["ExtEthCompatPayload"] = "ext_ethCompatPayload";
})(RenVMType || (RenVMType = {}));
