// send-crypto/build/module/common/apis/bitcoinDotCom.js
import axios from "axios";
import { fixUTXO, fixUTXOs, sortUTXOs } from "../../lib/utxo";
import { DEFAULT_TIMEOUT } from "./timeout";
const endpoint = (testnet) => testnet ? "https://trest.bitcoin.com/v2/" : "https://rest.bitcoin.com/v2/";
const fetchUTXO = (testnet) => async (txHash, vOut) => {
    const url = `${endpoint(testnet).replace(/\/$/, "")}/transaction/details/${txHash}`;
    const response = await axios.get(`${url}`, { timeout: DEFAULT_TIMEOUT });
    const utxo = response.data;
    return fixUTXO({
        txHash,
        amount: parseFloat(utxo.vout[vOut].value),
        // script_hex: utxo.scriptPubKey,
        vOut,
        confirmations: utxo.confirmations,
    }, 8);
};
const fetchConfirmations = (testnet) => async (txHash) => {
    const url = `${endpoint(testnet).replace(/\/$/, "")}/transaction/details/${txHash}`;
    const tx = (await axios.get(`${url}`, { timeout: DEFAULT_TIMEOUT })).data;
    return tx.confirmations;
};
const fetchUTXOs = (testnet) => async (address, confirmations) => {
    const url = `${endpoint(testnet).replace(/\/$/, "")}/address/utxo/${address}`;
    const response = await axios.get(url, { timeout: DEFAULT_TIMEOUT });
    return fixUTXOs(response.data.utxos.map(utxo => ({
        txHash: utxo.txid,
        amount: utxo.amount,
        // script_hex: utxo.scriptPubKey,
        vOut: utxo.vout,
        confirmations: utxo.confirmations,
    }))
        .filter(utxo => confirmations === 0 || utxo.confirmations >= confirmations), 8)
        .sort(sortUTXOs);
};
export const broadcastTransaction = (testnet) => async (txHex) => {
    const url = `${endpoint(testnet).replace(/\/$/, "")}/rawtransactions/sendRawTransaction`;
    const response = await axios.post(url, { "hexes": [txHex] }, { timeout: DEFAULT_TIMEOUT });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    return response.data[0];
};
export const BitcoinDotCom = {
    fetchUTXO,
    fetchUTXOs,
    fetchConfirmations,
    broadcastTransaction,
};
