// send-crypto/build/module/common/apis/blockchair.js
import axios from "axios";
import { sortUTXOs } from "../../lib/utxo";
import { DEFAULT_TIMEOUT } from "./timeout";
const fetchUTXO = (network) => async (txHash, vOut) => {
    const url = `https://api.blockchair.com/${network}/dashboards/transaction/${txHash}`;
    const response = (await axios.get(`${url}`, { timeout: DEFAULT_TIMEOUT })).data;
    if (!response.data[txHash]) {
        throw new Error(`Transaction not found.`);
    }
    const tx = response.data[txHash];
    const txBlock = tx.transaction.block_id;
    let latestBlock = response.context.state;
    if (latestBlock === 0) {
        const statsUrl = `https://api.blockchair.com/${network}/stats`;
        const statsResponse = (await axios.get(statsUrl)).data;
        latestBlock = statsResponse.data.blocks - 1;
    }
    const confirmations = txBlock === -1 ? 0 : Math.max(latestBlock - txBlock + 1, 0);
    return {
        txHash,
        vOut,
        amount: tx.outputs[vOut].value,
        confirmations,
    };
};
const fetchConfirmations = (network) => async (txHash) => {
    const url = `https://api.blockchair.com/${network}/dashboards/transaction/${txHash}`;
    const response = (await axios.get(`${url}`, { timeout: DEFAULT_TIMEOUT })).data;
    if (!response.data[txHash]) {
        throw new Error(`Transaction not found.`);
    }
    const txBlock = response.data[txHash].transaction.block_id;
    let latestBlock = response.context.state;
    if (latestBlock === 0) {
        const statsUrl = `https://api.blockchair.com/${network}/stats`;
        const statsResponse = (await axios.get(statsUrl)).data;
        latestBlock = statsResponse.data.blocks - 1;
    }
    return txBlock === -1 ? 0 : Math.max(latestBlock - txBlock + 1, 0);
};
const fetchUTXOs = (network) => async (address, confirmations) => {
    const url = `https://api.blockchair.com/${network}/dashboards/address/${address}?limit=0,100`;
    const response = (await axios.get(url, { timeout: DEFAULT_TIMEOUT })).data;
    let latestBlock = response.context.state;
    if (latestBlock === 0) {
        const statsUrl = `https://api.blockchair.com/${network}/stats`;
        const statsResponse = (await axios.get(statsUrl)).data;
        latestBlock = statsResponse.data.blocks - 1;
    }
    return response.data[address].utxo.map(utxo => ({
        txHash: utxo.transaction_hash,
        amount: utxo.value,
        vOut: utxo.index,
        confirmations: utxo.block_id === -1 ? 0 : latestBlock - utxo.block_id + 1,
    }))
        .filter(utxo => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(sortUTXOs);
};
export const broadcastTransaction = (network) => async (txHex) => {
    const url = `https://api.blockchair.com/${network}/push/transaction`;
    const response = await axios.post(url, { data: txHex }, { timeout: DEFAULT_TIMEOUT });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    return response.data.data.transaction_hash;
};
var Networks;
(function (Networks) {
    Networks["BITCOIN"] = "bitcoin";
    Networks["BITCOIN_CASH"] = "bitcoin-cash";
    Networks["LITECOIN"] = "litecoin";
    Networks["BITCOIN_SV"] = "bitcoin-sv";
    Networks["DOGECOIN"] = "dogecoin";
    Networks["DASH"] = "dash";
    Networks["GROESTLCOIN"] = "groestlcoin";
    Networks["BITCOIN_TESTNET"] = "bitcoin/testnet";
})(Networks || (Networks = {}));
export const Blockchair = {
    networks: Networks,
    fetchUTXO,
    fetchUTXOs,
    fetchConfirmations,
    broadcastTransaction,
};
;
;
