// send-crypto/build/module/common/apis/insight.js
import axios from "axios";
import https from "https";
import { fixUTXO, fixValue, sortUTXOs } from "../../lib/utxo";
import { DEFAULT_TIMEOUT } from "./timeout";
const fetchUTXOs = (insightURL) => async (address, confirmations) => {
    const url = `${insightURL.replace(/\/$/, "")}/addr/${address}/utxo`;
    const response = await axios.get(url, {
        // TODO: Remove when certificate is fixed.
        httpsAgent: new https.Agent({
            rejectUnauthorized: false
        }),
        timeout: DEFAULT_TIMEOUT,
    });
    const data = typeof response.data === "string" ? JSON.parse(response.data) : response.data;
    return data.map(utxo => ({
        txHash: utxo.txid,
        amount: utxo.satoshis || fixValue(utxo.amount, 8),
        // script_hex: utxo.scriptPubKey,
        vOut: utxo.vout,
        confirmations: utxo.confirmations,
    }))
        .filter(utxo => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(sortUTXOs);
};
const fetchConfirmations = (insightURL) => async (txHash) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/${txHash}`;
    const response = await axios.get(url, { timeout: DEFAULT_TIMEOUT });
    return response.data.confirmations;
};
const fetchUTXO = (insightURL) => async (txHash, vOut) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/${txHash}`;
    const tx = (await axios.get(url, { timeout: DEFAULT_TIMEOUT })).data;
    return fixUTXO({
        txHash,
        amount: parseFloat(tx.vout[vOut].value),
        vOut,
        confirmations: tx.confirmations,
    }, 8);
};
export const broadcastTransaction = (insightURL) => async (txHex) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/send`;
    const response = await axios.post(url, { rawtx: txHex }, { timeout: DEFAULT_TIMEOUT });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    return response.data.txid;
};
export const Insight = {
    fetchUTXO,
    fetchUTXOs,
    fetchConfirmations,
    broadcastTransaction,
};
