// @renproject/utils/build/module/chains.js
import { bchAddressFrom, bchAddressToHex, btcAddressFrom, btcAddressToHex, createBCHAddress, createBTCAddress, createZECAddress, getBitcoinCashUTXO, getBitcoinCashUTXOs, getBitcoinUTXO, getBitcoinUTXOs, getZcashUTXO, getZcashUTXOs, zecAddressFrom, zecAddressToHex, } from "@renproject/chains";
import { Chain, Tokens as CommonTokens, } from "@renproject/interfaces";
import { ripemd160, sha256 } from "ethereumjs-util";
import { Ox } from "./common";
import { parseRenContract } from "./renVMUtils";
// const hexOrBase64ToBuffer = (value: string | Buffer): Buffer =>
//     typeof value === "string" ?
//         value.slice(0, 2) === "0x" ?
//             Buffer.from(strip0x(value), "hex") :
//             Buffer.from(value, "base64") :
//         Buffer.from(value);
export const hash160 = (publicKey) => ripemd160(sha256(publicKey), false);
/**
 * Generate Gateway address for a cross-chain transfer's origin chain.
 */
export const generateAddress = (renContract, gHash, mpkh, isTestnet) => {
    const chain = parseRenContract(renContract).from;
    switch (chain) {
        case Chain.Bitcoin:
            return createBTCAddress(isTestnet, Ox(mpkh), gHash);
        case Chain.Zcash:
            return createZECAddress(isTestnet, Ox(mpkh), gHash);
        case Chain.BitcoinCash:
            return createBCHAddress(isTestnet, Ox(mpkh), gHash);
        default:
            throw new Error(`Unable to generate deposit address for chain ${chain}`);
    }
};
/**
 * Retrieves unspent deposits at the provided address.
 * An optional `confirmations` parameter limits UTXOs to ones with at least that
 * amount of confirmations.
 */
export const retrieveDeposits = async (_network, renContract, address, confirmations = 0) => {
    const chain = parseRenContract(renContract).from;
    switch (chain) {
        case Chain.Bitcoin:
            return (await getBitcoinUTXOs(_network)(address, confirmations)).map((utxo) => ({
                chain: Chain.Bitcoin,
                utxo,
            }));
        case Chain.Zcash:
            return (await getZcashUTXOs(_network)(address, confirmations)).map((utxo) => ({
                chain: Chain.Zcash,
                utxo,
            }));
        case Chain.BitcoinCash:
            // tslint:disable-next-line: no-unnecessary-type-assertion
            return (await getBitcoinCashUTXOs(_network)(address, confirmations)).map((utxo) => ({
                chain: Chain.BitcoinCash,
                utxo,
            }));
        default:
            throw new Error(`Unable to retrieve deposits for chain ${chain}`);
    }
};
/**
 * Returns the number of confirmations for the specified UTXO.
 */
export const retrieveUTXO = async (_network, chain, transaction) => {
    // tslint:disable-next-line: no-any
    const { txHash, vOut } = transaction;
    switch (chain) {
        case Chain.Bitcoin:
            return await getBitcoinUTXO(_network)(txHash, vOut);
        case Chain.Zcash:
            return await getZcashUTXO(_network)(txHash, vOut);
        case Chain.BitcoinCash:
            // tslint:disable-next-line: no-unnecessary-type-assertion
            return await getBitcoinCashUTXO(_network)(txHash, vOut);
        default:
            throw new Error(`Unable to retrieve deposits for chain ${chain}`);
    }
};
export const btcUtils = {
    getUTXOs: getBitcoinUTXOs,
    addressToHex: btcAddressToHex,
    addressFrom: btcAddressFrom,
};
export const zecUtils = {
    getUTXOs: getZcashUTXOs,
    addressToHex: zecAddressToHex,
    addressFrom: zecAddressFrom,
};
export const bchUtils = {
    getUTXOs: getBitcoinCashUTXOs,
    addressToHex: bchAddressToHex,
    addressFrom: bchAddressFrom,
};
export const Tokens = {
    // Bitcoin
    BTC: {
        ...CommonTokens.BTC,
        ...btcUtils,
    },
    // Zcash
    ZEC: {
        ...CommonTokens.ZEC,
        ...zecUtils,
    },
    // Bitcoin Cash
    BCH: {
        ...CommonTokens.BCH,
        ...bchUtils,
    },
};
