// @renproject/utils/build/module/parameterUtils.js
import { Chain } from "@renproject/interfaces";
import { parseRenContract, resolveInToken, resolveOutToken, resolveSendTo } from "./renVMUtils";
import { utils } from "./utils";
import { toBigNumber } from "./value";
/**
 * `resolveSendCall` simplifies the arguments required by RenJS by allowing
 * developers to pass in a non-contract address as the `sendTo` field.
 * This function checks if this is the case and makes the required changes to
 * the parameters;
 */
export const resolveSendCall = (network, params, lockAndMint) => {
    const { sendTo, sendAmount, suggestedAmount, txConfig, ...restOfParams } = params;
    const amount = sendAmount || suggestedAmount;
    // The contract call hasn't been provided - but `sendTo` has. We overwrite
    // the contract call with a simple adapter call.
    if (!sendTo) {
        throw new Error(`"sendTo" parameter must be provided.`);
    }
    lockAndMint = lockAndMint === undefined ? !!String(sendTo).match(/^(0x)[0-9a-fA-Z]{40}$/) : lockAndMint;
    const sendToken = lockAndMint ? resolveInToken(params.sendToken) : resolveOutToken(params.sendToken);
    const renContract = parseRenContract(sendToken);
    if (renContract.to === Chain.Ethereum) {
        // Mint
        return {
            ...restOfParams,
            suggestedAmount: amount,
            contractCalls: [{
                    sendTo: network.addresses.gateways.BasicAdapter.address,
                    contractFn: "mint",
                    contractParams: [
                        { type: "string", name: "_symbol", value: renContract.asset },
                        { type: "address", name: "_address", value: sendTo },
                    ],
                    txConfig,
                }],
        };
    }
    else {
        // Burn
        if (!amount) {
            throw new Error(`Send amount must be provided in order to send directly to an address.`);
        }
        const token = parseRenContract(sendToken).asset;
        const addressToHex = utils[token].addressToHex(sendTo);
        // const tokenAddress = await getTokenAddress(network, web3, sendToken);
        // const approve = {
        //     sendTo: tokenAddress,
        //     contractFn: "approve",
        //     contractParams: [
        //         { type: "address" as const, name: "spender", value: network.addresses.gateways.BasicAdapter.address },
        //         { type: "uint256" as const, name: "amount", value: toBigNumber(amount).toFixed() },
        //     ],
        //     txConfig,
        // };
        const gateway = network.addresses.gateways[`${token.toUpperCase()}Gateway`]._address;
        return {
            ...restOfParams,
            suggestedAmount: amount,
            contractCalls: [
                // approve,
                {
                    sendTo: gateway,
                    contractFn: "burn",
                    contractParams: [
                        { type: "bytes", name: "_to", value: addressToHex },
                        { type: "uint256", name: "_amount", value: toBigNumber(amount).toFixed() },
                    ],
                    // txConfig: { gas: 200000, ...txConfig },
                    txConfig,
                }
            ]
        };
    }
};
export const processLockAndMintParams = (_network, _params) => {
    const processors = [
        resolveSendTo({ isMint: true }),
    ];
    return processors.reduce((params, processor) => processor(params), _params);
};
export const processBurnAndReleaseParams = (_network, _params) => {
    const processors = [
        resolveSendTo({ isMint: false }),
    ];
    return processors.reduce((params, processor) => processor(params), _params);
};
