// @renproject/utils/build/module/value.js
/**
 * The `value` function converts between different cryptocurrency units.
 * See `value.spec.ts` for example usage.
 */
import BigNumber from "bignumber.js";
const convert = (valueIn, fromUnit, toUnit) => {
    return valueIn.multipliedBy(fromUnit).dividedBy(toUnit);
};
export const toBigNumber = (valueIn) => BigNumber.isBigNumber(valueIn) ? new BigNumber(valueIn) : new BigNumber(valueIn.toString());
export const toFixed = (input) => input.toFixed ? input.toFixed() : input.toString();
class BaseValue {
    constructor(unitTypes, valueIn, unit) {
        // tslint:disable-next-line: no-use-before-declare
        this.to = (unit) => convert(this.value, new BigNumber("1"), this.unitTypes[resolveUnit(unit)]);
        this.unitTypes = unitTypes;
        this.value = convert(toBigNumber(valueIn), this.unitTypes[unit], new BigNumber("1"));
    }
}
// BTC /////////////////////////////////////////////////////////////////////////
const BTCUnits = {
    sats: new BigNumber("0.00000001"),
    ubtc: new BigNumber("0.000001"),
    mbtc: new BigNumber("0.001"),
    btc: new BigNumber("1"),
};
class BTCValue extends BaseValue {
    constructor(valueIn, unit) {
        super(BTCUnits, valueIn, unit);
        this.sats = () => this.to("sats");
        this.btc = () => this.to("btc");
        this._smallest = () => this.to("sats");
        this._readable = () => this.to("btc");
    }
}
BTCValue.units = BTCUnits;
BTCValue.synonyms = [{ synonyms: ["bitcoin", "bitcoins", "tbtc"], unit: "btc" }]; // tbtc is used to indicate testnet btc
// BCH /////////////////////////////////////////////////////////////////////////
const BCHUnits = {
    sats: new BigNumber("0.00000001"),
    ubch: new BigNumber("0.000001"),
    mbch: new BigNumber("0.001"),
    bch: new BigNumber("1"),
};
class BCHValue extends BaseValue {
    constructor(valueIn, unit) {
        super(BCHUnits, valueIn, unit);
        this.sats = () => this.to("sats");
        this.bch = () => this.to("bch");
        this._smallest = () => this.to("sats");
        this._readable = () => this.to("bch");
    }
}
BCHValue.units = BCHUnits;
BCHValue.synonyms = [{ synonyms: ["bitcoin-cash", "bcash", "tbch"], unit: "btc" }];
// ZEC /////////////////////////////////////////////////////////////////////////
const ZECUnits = {
    zats: new BigNumber("0.00000001"),
    uzec: new BigNumber("0.000001"),
    mzec: new BigNumber("0.001"),
    zec: new BigNumber("1"),
};
class ZECValue extends BaseValue {
    constructor(valueIn, unit) {
        super(ZECUnits, valueIn, unit);
        this.zats = () => this.to("zats");
        this.zec = () => this.to("zec");
        this._smallest = () => this.to("zats");
        this._readable = () => this.to("zec");
    }
}
ZECValue.units = ZECUnits;
ZECValue.synonyms = [{ synonyms: ["zcash", "tzec", "taz"], unit: "zec" }];
// Sats ////////////////////////////////////////////////////////////////////////
const SatsUnits = {
    sats: new BigNumber("0.00000001"),
    bch: new BigNumber("1"),
    btc: new BigNumber("1"),
};
class SatsValue extends BaseValue {
    constructor(valueIn, unit) {
        super(SatsUnits, valueIn, unit);
        this.sats = () => this.to("sats");
        this.btc = () => this.to("btc");
        this.bch = () => this.to("bch");
        this._smallest = () => this.to("sats");
        this._readable = () => this.to("btc");
    }
}
SatsValue.units = SatsUnits;
SatsValue.synonyms = [{}];
// ETH /////////////////////////////////////////////////////////////////////////
const ETHUnits = {
    wei: new BigNumber("0.000000000000000001"),
    kwei: new BigNumber("0.000000000000001"),
    mwei: new BigNumber("0.000000000001"),
    gwei: new BigNumber("0.000000001"),
    finney: new BigNumber("0.001"),
    eth: new BigNumber("1"),
};
class ETHValue extends BaseValue {
    constructor(valueIn, unit) {
        super(ETHUnits, valueIn, unit);
        this.wei = () => this.to("wei");
        this.eth = () => this.to("eth");
        this._smallest = () => this.to("wei");
        this._readable = () => this.to("eth");
    }
}
ETHValue.units = ETHUnits;
ETHValue.synonyms = [{ synonyms: ["ethereum", "ether", "keth"], unit: "eth" }];
////////////////////////////////////////////////////////////////////////////////
// General /////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const valueClasses = [BTCValue, BCHValue, ZECValue, ETHValue];
const resolveUnit = (unitIn) => {
    const unit = unitIn.toLowerCase().replace(/µ/, "u");
    // Satoshis
    if (unit.slice(0, 3) === "sat") {
        return "sats";
    }
    for (const ValueClass of valueClasses) {
        // Check if the unit belongs to the class
        if (new Set(Object.keys(ValueClass.units)).has(unit)) {
            return unit;
        }
        // Handle synonyms
        for (const { synonyms, unit: synonymUnit } of ValueClass.synonyms) {
            if (new Set(synonyms).has(unit)) {
                return synonymUnit;
            }
        }
    }
    throw new Error(`Unknown unit ${unitIn}`);
};
export const value = (valueIn, unitIn) => {
    const unit = resolveUnit(unitIn);
    // Satoshis
    if (unit === "sats") {
        return new SatsValue(valueIn, "sats");
    }
    for (const ValueClass of valueClasses) {
        // Check if the unit belongs to the class
        if (new Set(Object.keys(ValueClass.units)).has(unit)) {
            return new ValueClass(valueIn, unit);
        }
    }
    throw new Error(`Unknown unit ${unitIn}`);
};
