// @renproject/interfaces/build/main/logger.js
"use strict";
// tslint:disable: no-any no-console
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const bignumber_js_1 = __importDefault(require("bignumber.js"));
var LogLevel;
(function (LogLevel) {
    LogLevel["Error"] = "error";
    LogLevel["Warn"] = "warn";
    LogLevel["Log"] = "log";
    LogLevel["Info"] = "info";
    LogLevel["Debug"] = "debug";
    LogLevel["Trace"] = "trace";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
const levelValue = (level) => {
    switch (level) {
        case LogLevel.Error: return 0;
        case LogLevel.Warn: return 1;
        case LogLevel.Log: return 2;
        case LogLevel.Info: return 3;
        case LogLevel.Debug: return 4;
        case LogLevel.Trace: return 5;
    }
};
const toString = (value) => {
    try {
        if (typeof value === "string") {
            return value;
        }
        if (bignumber_js_1.default.isBigNumber(value)) {
            return value.toFixed();
        }
        const seen = [];
        return JSON.stringify(value, (_key, val) => {
            if (val !== null && typeof val === "object") {
                if (seen.indexOf(val) >= 0) {
                    return;
                }
                seen.push(val);
            }
            return val;
        }, "    ");
    }
    catch (error) {
        try {
            return String(value);
        }
        catch (error) {
            return "";
        }
    }
};
/**
 * SimpleLogger is a implementation of the Logger interface that also supports
 * setting the log level.
 *
 * @export
 * @class SimpleLogger
 */
class SimpleLogger {
    constructor(level = LogLevel.Warn) {
        this.trace = (message, ...optionalParams) => {
            if (levelValue(this.level) >= levelValue(LogLevel.Trace)) {
                if (optionalParams.length) {
                    console.group(this.prefix(LogLevel.Trace) + toString(message));
                    console.trace(...optionalParams.map(toString));
                    console.groupEnd();
                }
                else {
                    console.trace(this.prefix(LogLevel.Trace) + toString(message), ...optionalParams.map(toString));
                }
            }
        };
        this.debug = (message, ...optionalParams) => {
            if (levelValue(this.level) >= levelValue(LogLevel.Debug)) {
                if (optionalParams.length) {
                    console.group(this.prefix(LogLevel.Debug) + toString(message));
                    console.debug(...optionalParams.map(toString));
                    console.groupEnd();
                }
                else {
                    console.debug(this.prefix(LogLevel.Debug) + toString(message), ...optionalParams.map(toString));
                }
            }
        };
        this.info = (message, ...optionalParams) => {
            if (levelValue(this.level) >= levelValue(LogLevel.Info)) {
                console.info(toString(message), ...optionalParams.map(toString));
            }
        };
        this.log = (message, ...optionalParams) => {
            if (levelValue(this.level) >= levelValue(LogLevel.Log)) {
                console.log(toString(message), ...optionalParams.map(toString));
            }
        };
        this.warn = (message, ...optionalParams) => {
            if (levelValue(this.level) >= levelValue(LogLevel.Warn)) {
                console.warn(toString(message), ...optionalParams.map(toString));
            }
        };
        this.error = (message, ...optionalParams) => {
            if (levelValue(this.level) >= levelValue(LogLevel.Error)) {
                console.error(toString(message), ...optionalParams.map(toString));
            }
        };
        this.prefix = (level) => `[RenJS][${level.toUpperCase()}] `;
        this.level = level;
    }
}
exports.SimpleLogger = SimpleLogger;
