// @renproject/interfaces/build/main/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The names of the messages sent between the Gateway iFrame and GatewayJS
 * library.
 */
var GatewayMessageType;
(function (GatewayMessageType) {
    GatewayMessageType["ToggleSettings"] = "toggleSettings";
    GatewayMessageType["Pause"] = "pause";
    GatewayMessageType["Resume"] = "resume";
    GatewayMessageType["Ready"] = "ready";
    GatewayMessageType["TransferDetails"] = "transferDetails";
    GatewayMessageType["Transfers"] = "Transfers";
    GatewayMessageType["GetTransfers"] = "GetTransfers";
    GatewayMessageType["Status"] = "status";
    GatewayMessageType["GetStatus"] = "getStatus";
    GatewayMessageType["TransferUpdated"] = "transferUpdated";
    GatewayMessageType["Cancel"] = "cancel";
    GatewayMessageType["Error"] = "error";
    GatewayMessageType["Done"] = "done";
    GatewayMessageType["SendEthereumTx"] = "sendEthereumTx";
    GatewayMessageType["GetEthereumTxStatus"] = "getEthereumTxStatus";
    GatewayMessageType["SendEthereumTxConfirmations"] = "sendEthereumTxConfirmations";
    GatewayMessageType["GetEthereumTxBurn"] = "getEthereumTxBurn";
    GatewayMessageType["FindMintTransaction"] = "findMintTransaction";
    GatewayMessageType["RequestNotificationPermission"] = "requestNotificationPermission";
    GatewayMessageType["ShowNotification"] = "showNotification";
    GatewayMessageType["Acknowledgement"] = "acknowledgement";
})(GatewayMessageType = exports.GatewayMessageType || (exports.GatewayMessageType = {}));
