// @renproject/interfaces/build/main/renVM.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RenNetwork;
(function (RenNetwork) {
    RenNetwork["Mainnet"] = "mainnet";
    RenNetwork["Chaosnet"] = "chaosnet";
    RenNetwork["Testnet"] = "testnet";
    RenNetwork["Devnet"] = "devnet";
    RenNetwork["Localnet"] = "localnet";
})(RenNetwork = exports.RenNetwork || (exports.RenNetwork = {}));
exports.RenNetworks = [RenNetwork.Mainnet, RenNetwork.Chaosnet, RenNetwork.Testnet, RenNetwork.Devnet, RenNetwork.Localnet];
exports.isRenNetwork = (maybeRenNetwork) => // tslint:disable-line: no-any
 exports.RenNetworks.indexOf(maybeRenNetwork) !== -1;
var Chain;
(function (Chain) {
    Chain["Bitcoin"] = "Btc";
    Chain["Ethereum"] = "Eth";
    Chain["Zcash"] = "Zec";
    Chain["BitcoinCash"] = "Bch";
})(Chain = exports.Chain || (exports.Chain = {}));
exports.Chains = [Chain.Bitcoin, Chain.Ethereum, Chain.Zcash, Chain.BitcoinCash];
exports.isChain = (maybeChain) => // tslint:disable-line: no-any
 exports.Chains.indexOf(maybeChain) !== -1;
var Asset;
(function (Asset) {
    Asset["BTC"] = "BTC";
    Asset["ZEC"] = "ZEC";
    Asset["ETH"] = "ETH";
    Asset["BCH"] = "BCH";
})(Asset = exports.Asset || (exports.Asset = {}));
exports.Assets = [Asset.BTC, Asset.ZEC, Asset.ETH, Asset.BCH];
exports.isAsset = (maybeAsset) => // tslint:disable-line: no-any
 exports.Assets.indexOf(maybeAsset) !== -1;
var RenContract;
(function (RenContract) {
    RenContract["Btc2Eth"] = "BTC0Btc2Eth";
    RenContract["Eth2Btc"] = "BTC0Eth2Btc";
    RenContract["Zec2Eth"] = "ZEC0Zec2Eth";
    RenContract["Eth2Zec"] = "ZEC0Eth2Zec";
    RenContract["Bch2Eth"] = "BCH0Bch2Eth";
    RenContract["Eth2Bch"] = "BCH0Eth2Bch";
})(RenContract = exports.RenContract || (exports.RenContract = {}));
exports.RenContracts = [RenContract.Btc2Eth, RenContract.Eth2Btc, RenContract.Zec2Eth, RenContract.Eth2Zec, RenContract.Bch2Eth, RenContract.Eth2Bch];
exports.isRenContract = (maybeRenContract) => // tslint:disable-line: no-any
 exports.RenContracts.indexOf(maybeRenContract) !== -1;
exports.Tokens = {
    // Bitcoin
    BTC: {
        Mint: RenContract.Btc2Eth,
        Btc2Eth: RenContract.Btc2Eth,
        Burn: RenContract.Eth2Btc,
        Eth2Btc: RenContract.Eth2Btc,
    },
    // Zcash
    ZEC: {
        Mint: RenContract.Zec2Eth,
        Zec2Eth: RenContract.Zec2Eth,
        Burn: RenContract.Eth2Zec,
        Eth2Zec: RenContract.Eth2Zec,
    },
    // Bitcoin Cash
    BCH: {
        Mint: RenContract.Bch2Eth,
        Bch2Eth: RenContract.Bch2Eth,
        Burn: RenContract.Eth2Bch,
        Eth2Bch: RenContract.Eth2Bch,
    },
};
var RenVMType;
(function (RenVMType) {
    RenVMType["TypeAddress"] = "address";
    RenVMType["TypeStr"] = "str";
    RenVMType["TypeB32"] = "b32";
    RenVMType["TypeB"] = "b";
    RenVMType["TypeI8"] = "i8";
    RenVMType["TypeI16"] = "i16";
    RenVMType["TypeI32"] = "i32";
    RenVMType["TypeI64"] = "i64";
    RenVMType["TypeI128"] = "i128";
    RenVMType["TypeI256"] = "i256";
    RenVMType["TypeU8"] = "u8";
    RenVMType["TypeU16"] = "u16";
    RenVMType["TypeU32"] = "u32";
    RenVMType["TypeU64"] = "u64";
    RenVMType["TypeU128"] = "u128";
    RenVMType["TypeU256"] = "u256";
    RenVMType["TypeRecord"] = "record";
    RenVMType["TypeList"] = "list";
    // Ext
    RenVMType["ExtTypeEthCompatAddress"] = "ext_ethCompatAddress";
    RenVMType["ExtTypeBtcCompatUTXO"] = "ext_btcCompatUTXO";
    RenVMType["ExtTypeBtcCompatUTXOs"] = "ext_btcCompatUTXOs";
    RenVMType["ExtTypeEthCompatTx"] = "ext_ethCompatTx";
    RenVMType["ExtEthCompatPayload"] = "ext_ethCompatPayload";
})(RenVMType = exports.RenVMType || (exports.RenVMType = {}));
