// @renproject/interfaces/build/main/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LockAndMintStatus;
(function (LockAndMintStatus) {
    LockAndMintStatus["Committed"] = "mint_committed";
    LockAndMintStatus["Deposited"] = "mint_deposited";
    LockAndMintStatus["Confirmed"] = "mint_confirmed";
    LockAndMintStatus["SubmittedToRenVM"] = "mint_submittedToRenVM";
    LockAndMintStatus["ReturnedFromRenVM"] = "mint_returnedFromRenVM";
    LockAndMintStatus["SubmittedToEthereum"] = "mint_submittedToEthereum";
    LockAndMintStatus["ConfirmedOnEthereum"] = "mint_confirmedOnEthereum";
})(LockAndMintStatus = exports.LockAndMintStatus || (exports.LockAndMintStatus = {}));
var BurnAndReleaseStatus;
(function (BurnAndReleaseStatus) {
    BurnAndReleaseStatus["Committed"] = "burn_committed";
    BurnAndReleaseStatus["SubmittedToEthereum"] = "burn_submittedToEthereum";
    BurnAndReleaseStatus["ConfirmedOnEthereum"] = "burn_confirmedOnEthereum";
    BurnAndReleaseStatus["SubmittedToRenVM"] = "burn_submittedToRenVM";
    BurnAndReleaseStatus["ReturnedFromRenVM"] = "burn_returnedFromRenVM";
    BurnAndReleaseStatus["NoBurnFound"] = "burn_noBurnFound";
})(BurnAndReleaseStatus = exports.BurnAndReleaseStatus || (exports.BurnAndReleaseStatus = {}));
var TxStatus;
(function (TxStatus) {
    // TxStatusNil is used for transactions that have not been seen, or are
    // otherwise unknown.
    TxStatus["TxStatusNil"] = "nil";
    // TxStatusConfirming is used for transactions that are currently waiting
    // for their underlying blockchain transactions to ne confirmed.
    TxStatus["TxStatusConfirming"] = "confirming";
    // TxStatusPending is used for transactions that are waiting for consensus
    // to be reached on when the transaction should be executed.
    TxStatus["TxStatusPending"] = "pending";
    // TxStatusExecuting is used for transactions that are currently being
    // executed.
    TxStatus["TxStatusExecuting"] = "executing";
    // TxStatusDone is used for transactions that have been successfully
    // executed.
    TxStatus["TxStatusDone"] = "done";
    // TxStatusReverted is used for transactions that were reverted during
    // execution.
    TxStatus["TxStatusReverted"] = "reverted";
})(TxStatus = exports.TxStatus || (exports.TxStatus = {}));
var EventType;
(function (EventType) {
    EventType["LockAndMint"] = "lockAndMint";
    EventType["BurnAndRelease"] = "burnAndRelease";
})(EventType = exports.EventType || (exports.EventType = {}));
