// send-crypto/build/main/common/apis/insight.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Insight = exports.broadcastTransaction = void 0;
const axios_1 = __importDefault(require("axios"));
const https_1 = __importDefault(require("https"));
const utxo_1 = require("../../lib/utxo");
const timeout_1 = require("./timeout");
const fetchUTXOs = (insightURL) => async (address, confirmations) => {
    const url = `${insightURL.replace(/\/$/, "")}/addr/${address}/utxo`;
    const response = await axios_1.default.get(url, {
        // TODO: Remove when certificate is fixed.
        httpsAgent: new https_1.default.Agent({
            rejectUnauthorized: false
        }),
        timeout: timeout_1.DEFAULT_TIMEOUT,
    });
    const data = typeof response.data === "string" ? JSON.parse(response.data) : response.data;
    return data.map(utxo => ({
        txHash: utxo.txid,
        amount: utxo.satoshis || utxo_1.fixValue(utxo.amount, 8),
        // script_hex: utxo.scriptPubKey,
        vOut: utxo.vout,
        confirmations: utxo.confirmations,
    }))
        .filter(utxo => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(utxo_1.sortUTXOs);
};
const fetchConfirmations = (insightURL) => async (txHash) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/${txHash}`;
    const response = await axios_1.default.get(url, { timeout: timeout_1.DEFAULT_TIMEOUT });
    return response.data.confirmations;
};
const fetchUTXO = (insightURL) => async (txHash, vOut) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/${txHash}`;
    const tx = (await axios_1.default.get(url, { timeout: timeout_1.DEFAULT_TIMEOUT })).data;
    return utxo_1.fixUTXO({
        txHash,
        amount: parseFloat(tx.vout[vOut].value),
        vOut,
        confirmations: tx.confirmations,
    }, 8);
};
exports.broadcastTransaction = (insightURL) => async (txHex) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/send`;
    const response = await axios_1.default.post(url, { rawtx: txHex }, { timeout: timeout_1.DEFAULT_TIMEOUT });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    return response.data.txid;
};
exports.Insight = {
    fetchUTXO,
    fetchUTXOs,
    fetchConfirmations,
    broadcastTransaction: exports.broadcastTransaction,
};
