// @renproject/utils/build/main/abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mintABITemplate = {
    constant: false,
    inputs: [
        {
            name: "_amount",
            type: "uint256"
        },
        {
            name: "_nHash",
            type: "bytes32"
        },
        {
            name: "_sig",
            type: "bytes"
        },
    ],
    outputs: [],
    payable: true,
    stateMutability: "payable",
    type: "function",
};
exports.payloadToABI = (methodName, payload) => {
    return [
        {
            name: methodName,
            type: "function",
            inputs: [
                ...(payload || []).map(value => ({ type: value.type, name: value.name })),
            ],
            outputs: [],
        }
    ];
};
exports.payloadToMintABI = (methodName, payload) => {
    return [
        Object.assign({}, mintABITemplate, { name: methodName, inputs: [
                ...(payload || []).map(value => ({ type: value.type, name: value.name })),
                ...(mintABITemplate.inputs ? mintABITemplate.inputs : []),
            ] })
    ];
};
