// @renproject/utils/build/main/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chains_1 = require("@renproject/chains");
const interfaces_1 = require("@renproject/interfaces");
const ethereumjs_util_1 = require("ethereumjs-util");
const common_1 = require("./common");
const renVMUtils_1 = require("./renVMUtils");
// const hexOrBase64ToBuffer = (value: string | Buffer): Buffer =>
//     typeof value === "string" ?
//         value.slice(0, 2) === "0x" ?
//             Buffer.from(strip0x(value), "hex") :
//             Buffer.from(value, "base64") :
//         Buffer.from(value);
exports.hash160 = (publicKey) => ethereumjs_util_1.ripemd160(ethereumjs_util_1.sha256(publicKey), false);
/**
 * Generate Gateway address for a cross-chain transfer's origin chain.
 */
exports.generateAddress = (renContract, gHash, mpkh, isTestnet) => {
    const chain = renVMUtils_1.parseRenContract(renContract).from;
    switch (chain) {
        case interfaces_1.Chain.Bitcoin:
            return chains_1.createBTCAddress(isTestnet, common_1.Ox(mpkh), gHash);
        case interfaces_1.Chain.Zcash:
            return chains_1.createZECAddress(isTestnet, common_1.Ox(mpkh), gHash);
        case interfaces_1.Chain.BitcoinCash:
            return chains_1.createBCHAddress(isTestnet, common_1.Ox(mpkh), gHash);
        default:
            throw new Error(`Unable to generate deposit address for chain ${chain}`);
    }
};
/**
 * Retrieves unspent deposits at the provided address.
 * An optional `confirmations` parameter limits UTXOs to ones with at least that
 * amount of confirmations.
 */
exports.retrieveDeposits = async (_network, renContract, address, confirmations = 0) => {
    const chain = renVMUtils_1.parseRenContract(renContract).from;
    switch (chain) {
        case interfaces_1.Chain.Bitcoin:
            return (await chains_1.getBitcoinUTXOs(_network)(address, confirmations)).map((utxo) => ({
                chain: interfaces_1.Chain.Bitcoin,
                utxo,
            }));
        case interfaces_1.Chain.Zcash:
            return (await chains_1.getZcashUTXOs(_network)(address, confirmations)).map((utxo) => ({
                chain: interfaces_1.Chain.Zcash,
                utxo,
            }));
        case interfaces_1.Chain.BitcoinCash:
            // tslint:disable-next-line: no-unnecessary-type-assertion
            return (await chains_1.getBitcoinCashUTXOs(_network)(address, confirmations)).map((utxo) => ({
                chain: interfaces_1.Chain.BitcoinCash,
                utxo,
            }));
        default:
            throw new Error(`Unable to retrieve deposits for chain ${chain}`);
    }
};
/**
 * Returns the number of confirmations for the specified UTXO.
 */
exports.retrieveUTXO = async (_network, chain, transaction) => {
    // tslint:disable-next-line: no-any
    const { txHash, vOut } = transaction;
    switch (chain) {
        case interfaces_1.Chain.Bitcoin:
            return await chains_1.getBitcoinUTXO(_network)(txHash, vOut);
        case interfaces_1.Chain.Zcash:
            return await chains_1.getZcashUTXO(_network)(txHash, vOut);
        case interfaces_1.Chain.BitcoinCash:
            // tslint:disable-next-line: no-unnecessary-type-assertion
            return await chains_1.getBitcoinCashUTXO(_network)(txHash, vOut);
        default:
            throw new Error(`Unable to retrieve deposits for chain ${chain}`);
    }
};
exports.btcUtils = {
    getUTXOs: chains_1.getBitcoinUTXOs,
    addressToHex: chains_1.btcAddressToHex,
    addressFrom: chains_1.btcAddressFrom,
};
exports.zecUtils = {
    getUTXOs: chains_1.getZcashUTXOs,
    addressToHex: chains_1.zecAddressToHex,
    addressFrom: chains_1.zecAddressFrom,
};
exports.bchUtils = {
    getUTXOs: chains_1.getBitcoinCashUTXOs,
    addressToHex: chains_1.bchAddressToHex,
    addressFrom: chains_1.bchAddressFrom,
};
exports.Tokens = {
    // Bitcoin
    BTC: Object.assign({}, interfaces_1.Tokens.BTC, exports.btcUtils),
    // Zcash
    ZEC: Object.assign({}, interfaces_1.Tokens.ZEC, exports.zecUtils),
    // Bitcoin Cash
    BCH: Object.assign({}, interfaces_1.Tokens.BCH, exports.bchUtils),
};
