// @renproject/utils/build/main/networks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const contracts_1 = require("@renproject/contracts");
exports.stringToNetwork = (network) => {
    if (typeof network === "string") {
        switch (network.toLowerCase()) {
            case "":
            case "mainnet":
                return contracts_1.renMainnet;
            case "chaosnet":
                return contracts_1.renChaosnet;
            case "testnet":
                return contracts_1.renTestnet;
            case "devnet":
                return contracts_1.renDevnet;
            // case "localnet":
            //     return localnet;
            default:
                throw new Error(`Unsupported network "${network}"`);
        }
    }
    else if (network === undefined || network === null) {
        return contracts_1.renMainnet;
    }
    else {
        return network;
    }
};
exports.NetworkDetails = {
    Mainnet: contracts_1.renMainnet,
    Chaosnet: contracts_1.renChaosnet,
    Testnet: contracts_1.renTestnet,
    Devnet: contracts_1.renDevnet,
    Localnet: contracts_1.renLocalnet,
    stringToNetwork: exports.stringToNetwork,
};
