// @renproject/utils/build/main/promiEvent.js
"use strict";
// Modified from web3-core-method
Object.defineProperty(exports, "__esModule", { value: true });
/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file PromiEvent.js
 * @author Fabian Vogelsteller <fabian@ethereum.org>, Samuel Furter <samuel@ethereum.org>
 * @date 2018
 */
// tslint:disable: no-any
const events_1 = require("events");
class InternalPromiEvent {
    /**
     * @constructor
     */
    constructor() {
        // tslint:disable-next-line: promise-must-complete
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
        this._cancelled = false;
        this.eventEmitter = new events_1.EventEmitter();
        return new Proxy(this, {
            get: this.proxyHandler
        });
    }
    /**
     * Proxy handler to call the promise or eventEmitter methods
     */
    proxyHandler(target, name) {
        if (name === "resolve" || name === "reject") {
            return target[name];
        }
        if (name === "then") {
            return target.promise.then.bind(target.promise);
        }
        if (name === "catch") {
            return target.promise.catch.bind(target.promise);
        }
        if (name === "_cancel") {
            // tslint:disable-next-line: no-object-mutation
            return () => { this._cancelled = true; };
        }
        if (name === "_isCancelled") {
            // tslint:disable-next-line: no-object-mutation
            return () => this._cancelled === true;
        }
        // tslint:disable-next-line: no-if-statement
        if (target.eventEmitter[name]) {
            return target.eventEmitter[name];
        }
    }
}
exports.InternalPromiEvent = InternalPromiEvent;
exports.newPromiEvent = () => new InternalPromiEvent();
