// @renproject/chains/build/main/bch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bchaddrjs_1 = require("bchaddrjs");
const bitcore_lib_cash_1 = require("bitcore-lib-cash");
const BCHHandler_1 = require("send-crypto/build/main/handlers/BCH/BCHHandler");
const btc_1 = require("./btc");
const common_1 = require("./common");
const hex_1 = require("./hex");
exports.createBCHAddress = common_1.createAddress(bitcore_lib_cash_1.Networks, bitcore_lib_cash_1.Opcode, bitcore_lib_cash_1.Script);
exports.getBitcoinCashUTXOs = ({ isTestnet }) => {
    return async (address, confirmations) => {
        return BCHHandler_1.getUTXOs(isTestnet, { address, confirmations });
    };
};
exports.getBitcoinCashConfirmations = ({ isTestnet, }) => {
    return async (txHash) => {
        return BCHHandler_1.getConfirmations(isTestnet, txHash);
    };
};
exports.getBitcoinCashUTXO = ({ isTestnet }) => {
    return async (txHash, vOut) => {
        return BCHHandler_1.getUTXO(isTestnet, txHash, vOut);
    };
};
exports.bchAddressToHex = (address) => hex_1.Ox(Buffer.from(address));
const isBCHAddress = (address, options) => {
    try {
        return options
            ? options.isTestnet
                ? bchaddrjs_1.isTestnetAddress(address)
                : bchaddrjs_1.isMainnetAddress(address)
            : bchaddrjs_1.isTestnetAddress(address) || bchaddrjs_1.isMainnetAddress(address);
    }
    catch (error) {
        return false;
    }
};
const bchTactics = {
    decoders: [
        (address) => Buffer.from(address),
        (address) => Buffer.from(address, "base64"),
        (address) => Buffer.from(hex_1.strip0x(address), "hex"),
    ],
    encoders: [(buffer) => bchaddrjs_1.toCashAddress(buffer.toString())],
};
exports.bchAddressFrom = btc_1.anyAddressFrom(isBCHAddress, bchTactics);
