// @renproject/chains/build/main/btc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bitcore_lib_1 = require("bitcore-lib");
const bs58_1 = require("bs58");
const BTCHandler_1 = require("send-crypto/build/main/handlers/BTC/BTCHandler");
const wallet_address_validator_1 = require("wallet-address-validator");
const common_1 = require("./common");
const hex_1 = require("./hex");
exports.createBTCAddress = common_1.createAddress(bitcore_lib_1.Networks, bitcore_lib_1.Opcode, bitcore_lib_1.Script);
exports.getBitcoinUTXOs = ({ isTestnet }) => {
    return async (address, confirmations) => {
        return BTCHandler_1.getUTXOs(isTestnet, { address, confirmations });
    };
};
exports.getBitcoinConfirmations = ({ isTestnet, }) => {
    return async (txHash) => {
        return BTCHandler_1.getConfirmations(isTestnet, txHash);
    };
};
exports.getBitcoinUTXO = ({ isTestnet }) => {
    return async (txHash, vOut) => {
        return BTCHandler_1.getUTXO(isTestnet, txHash, vOut);
    };
};
// export const btcAddressToHex = (address: string) => {
//     const addressBuffer = new Address(address).toBuffer();
//     // Concatenate checksum
//     return Ox(Buffer.concat([addressBuffer, Base58Check.checksum(addressBuffer)]));
// };
// export const btcAddressFrom = (address: string, encoding: "hex" | "base64") => {
//     // tslint:disable-next-line: no-any
//     return (Address as any)
//         .fromBuffer(Buffer.from(encoding === "hex" ? strip0x(address) : address, encoding).slice(0, -4))
//         .toString();
// };
exports.btcAddressToHex = (address) => hex_1.Ox(Buffer.from(address));
const isBTCAddress = (address) => wallet_address_validator_1.validate(address, "btc", "testnet") || wallet_address_validator_1.validate(address, "btc", "prod");
const btcTactics = {
    decoders: [
        (address) => Buffer.from(address),
        (address) => Buffer.from(address, "base64"),
        (address) => Buffer.from(hex_1.strip0x(address), "hex"),
    ],
    encoders: [
        (buffer) => bs58_1.encode(buffer),
        (buffer) => buffer.toString(),
    ],
};
exports.anyAddressFrom = (isAnyAddress, { encoders, decoders }) => (address) => {
    for (const encoder of encoders) {
        for (const decoder of decoders) {
            try {
                const encoded = encoder(decoder(address));
                if (isAnyAddress(encoded)) {
                    return encoded;
                }
            }
            catch (error) {
                // Ignore errors
            }
        }
    }
    return address;
};
exports.btcAddressFrom = exports.anyAddressFrom(isBTCAddress, btcTactics);
