// @renproject/chains/build/main/zec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const bitcore_lib_zcash_1 = require("bitcore-lib-zcash");
const base58check_1 = __importDefault(require("bitcore-lib-zcash/lib/encoding/base58check"));
const bs58_1 = require("bs58");
const ZECHandler_1 = require("send-crypto/build/main/handlers/ZEC/ZECHandler");
const wallet_address_validator_1 = require("wallet-address-validator");
const btc_1 = require("./btc");
const common_1 = require("./common");
const hex_1 = require("./hex");
exports.createZECAddress = common_1.createAddress(bitcore_lib_zcash_1.Networks, bitcore_lib_zcash_1.Opcode, bitcore_lib_zcash_1.Script);
exports.getZcashUTXOs = ({ isTestnet }) => {
    return async (address, confirmations) => {
        return ZECHandler_1.getUTXOs(isTestnet, { address, confirmations });
    };
};
exports.getZcashConfirmations = ({ isTestnet, }) => {
    return async (txHash) => {
        return ZECHandler_1.getConfirmations(isTestnet, txHash);
    };
};
exports.getZcashUTXO = ({ isTestnet }) => {
    return async (txHash, vOut) => {
        return ZECHandler_1.getUTXO(isTestnet, txHash, vOut);
    };
};
// ZCash shielded addresses (starting with 'z') aren't supported yet.
// export const zecAddressToHex = (address: string) => Ox(Buffer.from(address));
exports.zecAddressToHex = (address) => {
    const addressBuffer = new bitcore_lib_zcash_1.Address(address).toBuffer();
    // Concatenate checksum
    return hex_1.Ox(Buffer.concat([addressBuffer, base58check_1.default.checksum(addressBuffer)]));
};
const isZECAddress = (address) => wallet_address_validator_1.validate(address, "zec", "testnet") || wallet_address_validator_1.validate(address, "zec", "prod");
const zecTactics = {
    decoders: [
        (address) => Buffer.from(address),
        (address) => Buffer.from(address, "base64"),
        (address) => Buffer.from(hex_1.strip0x(address), "hex"),
    ],
    encoders: [
        (buffer) => buffer.toString(),
        (buffer) => bs58_1.encode(buffer),
    ],
};
exports.zecAddressFrom = btc_1.anyAddressFrom(isZECAddress, zecTactics);
